File: core/pricing.py
"""
Pricing engine.

Implements the pricing rule from the architecture this project is based
on: bid at exactly 10% below an estimated market price, never below a hard
COGS + minimum-margin floor. The floor check is a hard invariant -- the
Reviewer stage (core.reviewer) re-verifies it independently before any bid
can be approved, so a Strategist/Writer bug can never ship an underpriced
bid on its own.
"""
from __future__ import annotations

from dataclasses import dataclass

from core.models import ManufacturingDomain

DISCOUNT_VS_MARKET = 0.90  # bid = 90% of estimated market price
MIN_MARGIN_PCT = 0.35      # never bid below cogs * (1 + this)

ALUMINUM_USD_PER_KG = 3.15
ELECTRICITY_USD_PER_KWH = 0.12

# Rough base-rate table (USD) used to estimate "what a traditional shop
# would charge" for a given domain before volume/complexity adjustment.
# These are illustrative defaults, not a live market feed -- swap in a
# real pricing model / comps API when one exists.
_BASE_MARKET_RATE = {
    ManufacturingDomain.CAD_MECHANICAL: 1800.0,
    ManufacturingDomain.PCB_ELECTRONICS: 2600.0,
    ManufacturingDomain.ENCLOSURE_DESIGN: 1500.0,
    ManufacturingDomain.PROTOTYPING_3D_PRINT: 900.0,
    ManufacturingDomain.DFM_INJECTION_MOLDING: 6500.0,
    ManufacturingDomain.SHEET_METAL: 3200.0,
    ManufacturingDomain.FULL_NPD_TURNKEY: 12000.0,
    ManufacturingDomain.UNKNOWN: 1200.0,
}


@dataclass
class PriceQuote:
    market_price_usd: float
    bid_price_usd: float
    cogs_usd: float
    margin_pct: float
    floor_violation: bool


def estimate_market_price(domain: ManufacturingDomain, target_volume: int | None, stated_budget_usd: float | None) -> float:
    base = _BASE_MARKET_RATE.get(domain, 1200.0)
    if target_volume and target_volume > 1:
        # volume adds tooling/setup amortization, with diminishing per-unit weight
        volume_factor = min(target_volume, 100_000) ** 0.35
        base += volume_factor * 12.0
    if stated_budget_usd:
        # anchor loosely to what the client already signaled, without
        # letting a lowball ask collapse the estimate entirely
        base = (base * 0.6) + (stated_budget_usd * 0.4)
    return round(base, 2)


def estimate_cogs(domain: ManufacturingDomain, target_volume: int | None) -> float:
    units = max(target_volume or 1, 1)
    est_mass_kg = 0.35 if domain in (
        ManufacturingDomain.ENCLOSURE_DESIGN,
        ManufacturingDomain.PROTOTYPING_3D_PRINT,
        ManufacturingDomain.DFM_INJECTION_MOLDING,
        ManufacturingDomain.SHEET_METAL,
    ) else 0.05
    material_cost = units * est_mass_kg * ALUMINUM_USD_PER_KG
    est_kwh_per_unit = 0.4
    power_cost = units * est_kwh_per_unit * ELECTRICITY_USD_PER_KWH
    # cap the modeled run to a representative production batch so a
    # 100k-unit RFQ doesn't dominate the single-lead cost estimate
    modeled_units = min(units, 500)
    scale = modeled_units / units
    return round((material_cost + power_cost) * scale + 40.0, 2)  # +40 flat engineering setup


def quote(domain: ManufacturingDomain, target_volume: int | None, stated_budget_usd: float | None) -> PriceQuote:
    market = estimate_market_price(domain, target_volume, stated_budget_usd)
    bid = round(market * DISCOUNT_VS_MARKET, 2)
    cogs = estimate_cogs(domain, target_volume)
    floor = cogs * (1 + MIN_MARGIN_PCT)
    floor_violation = bid < floor
    if floor_violation:
        bid = round(floor, 2)
    margin_pct = ((bid - cogs) / bid * 100.0) if bid else 0.0
    return PriceQuote(
        market_price_usd=market,
        bid_price_usd=bid,
        cogs_usd=cogs,
        margin_pct=margin_pct,
        floor_violation=floor_violation,
    )

File: iechm-leadgen-engine-c/src/pricing.c
#include "pricing.h"
#include <math.h>

const double MIN_MARGIN_PCT = 0.35;
static const double DISCOUNT_VS_MARKET = 0.90;
static const double ALUMINUM_USD_PER_KG = 3.15;
static const double ELECTRICITY_USD_PER_KWH = 0.12;

static double round2(double v) { return round(v * 100.0) / 100.0; }

static double base_market_rate(ManufacturingDomain d) {
    switch (d) {
        case DOM_CAD_MECHANICAL: return 1800.0;
        case DOM_PCB_ELECTRONICS: return 2600.0;
        case DOM_ENCLOSURE_DESIGN: return 1500.0;
        case DOM_PROTOTYPING_3D_PRINT: return 900.0;
        case DOM_DFM_INJECTION_MOLDING: return 6500.0;
        case DOM_SHEET_METAL: return 3200.0;
        case DOM_FULL_NPD_TURNKEY: return 12000.0;
        case DOM_UNKNOWN: default: return 1200.0;
    }
}

static double estimate_market_price(ManufacturingDomain domain, bool has_volume, int target_volume,
                                     bool has_budget, double stated_budget_usd) {
    double base = base_market_rate(domain);
    if (has_volume && target_volume > 1) {
        double capped = target_volume > 100000 ? 100000.0 : (double) target_volume;
        double volume_factor = pow(capped, 0.35);
        base += volume_factor * 12.0;
    }
    if (has_budget) {
        base = (base * 0.6) + (stated_budget_usd * 0.4);
    }
    return round2(base);
}

static double estimate_cogs(ManufacturingDomain domain, bool has_volume, int target_volume) {
    int units = has_volume ? target_volume : 1;
    if (units < 1) units = 1;
    bool heavier = (domain == DOM_ENCLOSURE_DESIGN || domain == DOM_PROTOTYPING_3D_PRINT ||
                    domain == DOM_DFM_INJECTION_MOLDING || domain == DOM_SHEET_METAL);
    double est_mass_kg = heavier ? 0.35 : 0.05;
    double material_cost = units * est_mass_kg * ALUMINUM_USD_PER_KG;
    double est_kwh_per_unit = 0.4;
    double power_cost = units * est_kwh_per_unit * ELECTRICITY_USD_PER_KWH;
    int modeled_units = units < 500 ? units : 500;
    double scale = (double) modeled_units / (double) units;
    return round2((material_cost + power_cost) * scale + 40.0);
}

PriceQuote price_quote(ManufacturingDomain domain, bool has_volume, int target_volume,
                        bool has_budget, double stated_budget_usd) {
    PriceQuote q;
    q.market_price_usd = estimate_market_price(domain, has_volume, target_volume, has_budget, stated_budget_usd);
    q.bid_price_usd = round2(q.market_price_usd * DISCOUNT_VS_MARKET);
    q.cogs_usd = estimate_cogs(domain, has_volume, target_volume);
    double floor = q.cogs_usd * (1.0 + MIN_MARGIN_PCT);
    q.floor_violation = q.bid_price_usd < floor;
    if (q.floor_violation) q.bid_price_usd = round2(floor);
    q.margin_pct = q.bid_price_usd != 0.0 ? (q.bid_price_usd - q.cogs_usd) / q.bid_price_usd * 100.0 : 0.0;
    return q;
}
